src/source/Ice/IceAgentStateMachine.c

```diff
@@ -200,20 +200,42 @@ STATUS executeNewIceAgentState(UINT64 customData, UINT64 time)
     UNUSED_PARAM(time);
     STATUS retStatus = STATUS_SUCCESS;
     PIceAgent pIceAgent = (PIceAgent) customData;
+    //assumes this is called while holding pIceAgent->lock
+    BOOL locked = TRUE;
 
     CHK(pIceAgent != NULL, STATUS_NULL_ARG);
     // return early if we are already in ICE_AGENT_STATE_GATHERING
     CHK(pIceAgent->iceAgentState != ICE_AGENT_STATE_NEW, retStatus);
 
     pIceAgent->iceAgentState = ICE_AGENT_STATE_NEW;
 
+    //unlocking here to make thread sanitizer happy
```

Comment:
Ah, an interesting case! Nice job TSAN! Here is what's happening. The NEW state is only stepped into from the thread (the application thread) that calls peer connection create. The path leads to creating a new IceAgent which creates a state machine, sets things up and steps it. Each IceAgent object encapsulates its own state machine. We have a higher level stepping function of the state machine that does a few things and acquires a lock before stepping the state machine. Normally, the timer queue thread will get scheduled after this function returns and unlocks the ice agent lock. The timer queue timer that gets set is a multiple time firing timer, however, it's scheduled to start immediately (the second param is 0). The TSAN schedules the timer callback before the executeNewIceAgentState has a chance to return an unlock the lock.

I believe this needs to be properly fixed.

@chehefen has a massive rework of the Ice agent code. Let me pull him in so we can address this race condition properly.

Reply:
good 
thats exactly how i understood what was going on
and yes it's a stop gap to make thread sanitizer happy and avoid a _potential_ deadlock 
should be fixed properly
i was thinking of separate locks for timer queue and executor
queue gets locked for put and get operations

@chehefen what's your approach?

Reply:
I schedule the timer in iceAgentStartAgent: https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/322/files#diff-011ad0584af4444f14cb6ba52959c0a4R435-R444 instead. Since in the PR I got rid of the gathering state, and wait for credential state, state transition would start with check connection state, and there is no point checking connection until we receive the user name and password in iceAgentStartAgent. I also got rid of stepStateMachine call in createIceAgent for the same reason.